### PR TITLE
Add Unittests for ODBC

### PR DESCRIPTION
**Pull request by maiadegraaf**
Rewrote the following ODBC tests from psqlodbc so that they no longer require comparing stdout.
- alter
- bind_col
- catalog_functions
- col_attribute
- connect
- diagnostics
- row_wise_fetching
- select

@taniabogatsch

### CODE REVIEW

**Comment by Mytherin** [comment]
Thanks for the PR! LGTM. These test designs will make it much easier to run the ODBC tests.

It seems that not all ODBC tests have run as part of this PR however. It looks like `OSX.yml` and `LinuxRelease.yml` incorrectly have `tools/odbc` in the `paths-ignore`. @carlopi could you perhaps have a look at that and help fix that?

File: tools/odbc/test/tests/alter.cpp [hunk 1]
```diff
@@ -0,0 +1,51 @@
+#include "../common.h"
+
+using namespace odbc_test;
+
+TEST_CASE("Alter", "[odbc]") {
```
Comment (taniabogatsch) [inline tools/odbc/test/tests/alter.cpp:5]: Nit: Maybe elaborate a bit more on the test names, e.g., `"Test ALTER TABLE statement"`

File: tools/odbc/test/tests/alter.cpp [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+	// Create a table to test with
+	ExecuteCmdAndCheckODBC("SQLExecDirect (CREATE TABLE)", SQLExecDirect, hstmt,
+	                       ConvertToSQLCHAR("CREATE TABLE testtbl(t varchar(40))"), SQL_NTS);
+
+	// A simple query against the table, fetch column info
+	ExecuteCmdAndCheckODBC("SQLExecDirect (SELECT)", SQLExecDirect, hstmt, ConvertToSQLCHAR("SELECT * FROM testtbl"),
+	                       SQL_NTS);
+
+	// Get column metadata
```
Comment (taniabogatsch) [inline tools/odbc/test/tests/alter.cpp:24]: Nit: Get the number of columns

File: tools/odbc/test/tests/alter.cpp [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+	// Alter the table
+	ExecuteCmdAndCheckODBC("SQLExecDirect (ALTER TABLE)", SQLExecDirect, hstmt,
+	                       ConvertToSQLCHAR("ALTER TABLE testtbl ALTER t SET DATA TYPE int"), SQL_NTS);
+
+	// Rerun the query to check if the metadata was updated
+	ExecuteCmdAndCheckODBC("SQLExecDirect (SELECT)", SQLExecDirect, hstmt, ConvertToSQLCHAR("SELECT * FROM testtbl"),
+	                       SQL_NTS);
+
+	// Get column metadata
```
Comment (taniabogatsch) [inline tools/odbc/test/tests/alter.cpp:40]: See above, maybe `num_cols_updated`.

File: tools/odbc/test/tests/bind_col.cpp [hunk 4]
```diff
@@ -0,0 +1,85 @@
+#include "../common.h"
+
+#include <iostream>
```
Comment (taniabogatsch) [inline tools/odbc/test/tests/bind_col.cpp:3]: Nit: Still necessary?

File: tools/odbc/test/tests/bind_col.cpp [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+
+using namespace odbc_test;
+
+TEST_CASE("bind_col", "[odbc") {
```
Comment (taniabogatsch) [inline tools/odbc/test/tests/bind_col.cpp:7]: `"Test binding columns to application data buffers"`

File: tools/odbc/test/tests/bind_col.cpp [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+
+	// Execute the query
+	ExecuteCmdAndCheckODBC("SQLExecDirect (HSTMT)", SQLExecDirect, hstmt,
+	                       ConvertToSQLCHAR("SELECT id, 'foo' || id FROM generate_series(1, 10) id(id)"), SQL_NTS);
+
+	SQLINTEGER id = 0;
+	SQLINTEGER foo = 0;
+	SQLINTEGER row_num = 0;
+	bool incr_id = true;
+	bool incr_foo = true;
```
Comment (taniabogatsch) [inline tools/odbc/test/tests/bind_col.cpp:39]: Might be easier to use an array containing the expected output of each column.

File: tools/odbc/test/tests/connect.cpp [hunk 7]
```diff
@@ -0,0 +1,17 @@
+#include "../common.h"
+#include <iostream>
```
Comment (taniabogatsch) [inline tools/odbc/test/tests/connect.cpp:2]: Nit: necessary?

File: tools/odbc/test/tests/bools_as_char.cpp [hunk 8]
```diff
@@ -0,0 +1,27 @@
+#include "../common.h"
```
Comment (taniabogatsch) [inline tools/odbc/test/tests/bools_as_char.cpp:1]: Remove or add this to the top and uncomment the code:
```cpp
// FIXME: add the actual test body
return;
```

File: tools/odbc/test/tests/row_wise_fetching.cpp [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+	OrderInfo order_info[ROW_ARRAY_SIZE];
+	SQLULEN rows_fetched;
+	SQLUSMALLINT row_array_status[ROW_ARRAY_SIZE];
+	SQLULEN order_info_size = sizeof(order_info);
+	ExecuteCmdAndCheckODBC("SQLSetStmtAttr", SQLSetStmtAttr, hstmt, SQL_ATTR_ROW_ARRAY_SIZE,
+	                       reinterpret_cast<SQLPOINTER>(ROW_ARRAY_SIZE), 0);
+}
+
+TEST_CASE("row_wise_fetching", "[odbc") {
+	SQLRETURN ret;
```
Comment (taniabogatsch) [inline tools/odbc/test/tests/row_wise_fetching.cpp:28]: Same as for `bools_as_char.cpp`.

File: tools/odbc/test/tests/row_wise_fetching.cpp [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+static void TestMicrosoftExample(HSTMT &hstmt) {
+	OrderInfo order_info[ROW_ARRAY_SIZE];
+	SQLULEN rows_fetched;
+	SQLUSMALLINT row_array_status[ROW_ARRAY_SIZE];
+	SQLULEN order_info_size = sizeof(order_info);
+	ExecuteCmdAndCheckODBC("SQLSetStmtAttr", SQLSetStmtAttr, hstmt, SQL_ATTR_ROW_ARRAY_SIZE,
+	                       reinterpret_cast<SQLPOINTER>(ROW_ARRAY_SIZE), 0);
+}
+
+TEST_CASE("row_wise_fetching", "[odbc") {
```
Comment (taniabogatsch) [inline tools/odbc/test/tests/row_wise_fetching.cpp:27]: I think that a lot of these should have a closing bracket `]`. 👀

**Review by taniabogatsch — CHANGES_REQUESTED** [review]
Thanks for the PR; looks pretty good already! I added our peer review comments below, and we should get together later to look at the four missing test files. 
`catalog_functions`
`col_attribute`
`diagnostics`
`select`

**Review by taniabogatsch — CHANGES_REQUESTED** [review]
Here goes the review of the tests we didn't cover in the previous round. :) I also added a few additional thoughts below.

I think extending the ODBC `README.md` with some of the points we discussed while looking into this PR will be useful. :) Maybe we should also emphasize double-checking and looking up things in the official documentation [here](https://learn.microsoft.com/en-us/sql/connect/odbc/microsoft-odbc-driver-for-sql-server?view=sql-server-ver16)...

And maybe @Mytherin has thoughts on this, but it might be helpful to create a subdirectory `odbc/test/data` containing CSVs (or another format) with the expected output. What we saw a lot in this PR are long arrays containing strings, e.g., the code below.
```cpp
...
	expected_metadata.push_back({"MINIMUM_SCALE", SQL_SMALLINT});
	expected_data.emplace_back("-1");
	expected_metadata.push_back({"MAXIMUM_SCALE", SQL_SMALLINT});
	expected_data.emplace_back("-1");
	expected_metadata.push_back({"SQL_DATA_TYPE", SQL_SMALLINT});
	expected_data.emplace_back("12");
	expected_metadata.push_back({"SQL_DATETIME_SUB", SQL_SMALLINT});
	expected_data.emplace_back("-1");
	expected_metadata.push_back({"NUM_PREC_RADIX", SQL_INTEGER});
	expected_data.emplace_back("-1");
	expected_metadata.push_back({"INTERVAL_PRECISION", SQL_SMALLINT});
	expected_data.emplace_back("-1");
...
```

**Comment by taniabogatsch** [comment]
And as a last comment, to properly run the CI. We need to include the folder in the respective `.yml` files. As a reference, they are included in `Windows.yml` (see below). We must do that for `OSX.yml`, `LinuxRelease.yml`, and possibly `Main.yml`. @maiadegraaf let me know if you need help with that. 
```
   push:
[...]
    paths-ignore:
      - '**.md'
      - 'tools/**'
      - '!tools/odbc/**'
[...]
  pull_request:
    paths-ignore:
      - '**.md'
      - 'tools/**'
      - '!tools/odbc/**'
[...]
```

**Comment by carlopi** [comment]
> Thanks for the PR! LGTM. These test designs will make it much easier to run the ODBC tests.
> 
> It seems that not all ODBC tests have run as part of this PR however. It looks like `OSX.yml` and `LinuxRelease.yml` incorrectly have `tools/odbc` in the `paths-ignore`. @carlopi could you perhaps have a look at that and help fix that?

Thanks for pointing this out, checked with @maiadegraaf and @taniabogatsch, a commit addressing this is part of the PR.

File: .github/workflows/Regression.yml [hunk 11]
```diff
@@ -95,12 +95,6 @@ jobs:
       run: |
         python scripts/regression_test_runner.py --old=duckdb/build/release/benchmark/benchmark_runner --new=build/release/benchmark/benchmark_runner --benchmarks=.github/regression/tpch.csv --verbose --threads=2
 
-    - name: Regression Test TPCH-PARQUET
-      if: always()
-      shell: bash
-      run: |
-        python scripts/regression_test_runner.py --old=duckdb/build/release/benchmark/benchmark_runner --new=build/release/benchmark/benchmark_runner --benchmarks=.github/regression/tpch_parquet.csv --verbose --threads=2
-
```
Comment (taniabogatsch) [inline .github/workflows/Regression.yml:103]: Maybe I missed something, but why are we removing this?

File: tools/odbc/test/CMakeLists.txt [hunk 12]
```diff
@@ -2,5 +2,20 @@ include_directories(../include)
...
+  tests/basic_usage.cpp
+  tests/bind_col.cpp
+  tests/bools_as_char.cpp
+  tests/catalog_functions.cpp
+  tests/col_attribute.cpp
+  tests/connect.cpp
+  tests/diagnostics.cpp
+  tests/select.cpp
+  tests/row_wise_fetching.cpp
+  # tests/tmp.cpp
```
Comment (taniabogatsch) [inline tools/odbc/test/CMakeLists.txt:18]: Nit: remove.

File: tools/odbc/test/tests/bind_col.cpp [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+	EXECUTE_AND_CHECK("SQLBindCol (HSTMT)", SQLBindCol, hstmt, 2, SQL_C_CHAR, &char_value, sizeof(char_value),
+	                  &ind_char_value);
+
+	// Execute the query
+	EXECUTE_AND_CHECK("SQLExecDirect (HSTMT)", SQLExecDirect, hstmt,
+	                  ConvertToSQLCHAR("SELECT id, 'foo' || id FROM generate_series(1, 10) id(id)"), SQL_NTS);
+
+	SQLINTEGER row_num = 0;
+	std::array<int, 10> id = {1, 2, 3, 4, 5, 6, 7, 7, 7, 7};
+	std::array<std::string, 10> foo = {"foo1", "foo2", "foo3", "foo3", "foo3", "foo6", "foo7", "foo7", "foo7", "foo10"};
```
Comment (taniabogatsch) [inline tools/odbc/test/tests/bind_col.cpp:35]: Very nice!

**Review by taniabogatsch — APPROVED** [review]
Looks very neat now! From my side, this is good to go after resolving my last few comments. 😄

File: .github/workflows/Regression.yml (hunk 11 above)
Comment (maiadegraaf) [reply .github/workflows/Regression.yml:103]: I don't think I removed this but maybe it didn't get merged correctly with feature at some point?

**Comment by Mytherin** [comment]
Thanks! Looks great

**Pull request merged**